src/ui/preferences-window/tabs/GeneralTab.swift

```diff
@@ -1,4 +1,5 @@
 import Cocoa
+import LoginServiceKit
```

Comment:
This change puzzles me. It seems you switching from existing working code to a third-party [library which uses the same Apple APIs](https://github.com/Clipy/LoginServiceKit/blob/master/Lib/LoginServiceKit/LoginServiceKit.swift).

What was the motivation to do this?

Also, how can you be using `LoginServiceKit` without having it listed in the `Podfile`?

Reply:
I used the Swift Package Manager to include the dependency. And I think, all other dependencies should use the Swift Package Manager too (in the future).

And I used this dependency to solve an issue in the code https://github.com/lwouis/alt-tab-macos/blob/master/src/ui/preferences-window/tabs/GeneralTab.swift#L32 for now.

This API can be removed soon by Apple. But I didn't want to remove any functionality through this PR.

Reply:
> And I used this dependency to solve an issue in the code

As I said above, the library you introduced uses the same Apple APIs as AltTab uses currently. Thus your change doesn't improve on the issue you perceived

Regarding the deprecation, nothing can be done for now. Apple marked it as deprecated, but the alternative introduced since are different and not really replacement. I've spent hours understand the various APIs, their history, their pros and cons before landing on the current implementation.

When the day comes (if it comes) when Apple removes it, then we will have ample time to find a creative solution to replace it. For now this API is the best implementation in my opinion.

But again, here we are discussing a library that uses the same exact deprecated APIs, so it's literally replacing the API calls by a third-party lib that wraps the same API calls.

Reply:
Hey, I already said that with `This API can be removed soon by Apple.`.

But the current implementation just doesn't work and I didn't want to invest much time there, since LoginServiceKit also runs with macOS 11.5 without problems.

![Screenshot 2021-09-10 at 15 51 00](https://user-images.githubusercontent.com/92227/132863912-48c0ed57-2f25-4faa-bd45-b7a0f2c3a783.png)

Reply:
Again, I know we're talking about a long deprecated API here. Which is why I don't want to waste more time here.

And the current implementation has a bug, the external library does not have that, what is the idea here? Fix the current implementation? Is there a reason why the few lines must not come from a dependency?

We'll have to remove that soon anyway.

Reply:
> And the current implementation has a bug, the external library does not have that, what is the idea here? Fix the current implementation? Is there a reason why the few lines must not come from a dependency?

I think it's interesting to understand the root cause of the bug you found on macOS 11. The API is still present on that system, so why is it crashing? What's the difference between the code-path in the implementation vs the code-path in the library code? It may reveal another issue actually. I'm interested in understanding why suddenly an issue appeared in an area of code that has been stable and working for a long time now.

> We'll have to remove that soon anyway.

It's been deprecated for 6 years. That puts the "soon" in perspective I think.

Reply:
Couple things about this by the way, I found that this wasn't building for me on both master, and this pr
master on Xcode 13.1:
![image](https://user-images.githubusercontent.com/4915682/140657007-5ac16f03-bd66-4ac6-9d95-f05dc3de4ddc.png)

On master this is fixable by adding a few nil handlers. But on this branch I couldn't fix it due to the reliance on the package. imo, this package seems unnecessary, I found I was having a crash on startup, and so I commented out `kLSSharedFileListItemBeforeFirst` and then things worked as expected. I don't see this line in LoginServiceKit, so I think it seems reasonable to remove this line @lwouis ? And that way you don't need to download a whole new package to handle all this

Reply:
> I found that this wasn't building for me on both master, and this pr master on Xcode 13.1:

Do you know why there was a change (from compiling to no-longer compiling)? What's the new error with `nil` here?

> I don't see this line in LoginServiceKit, so I think it seems reasonable to remove this line @lwouis ?

Do you mean this line?

```
let _ = LSSharedFileListInsertItemURL(loginItems, kLSSharedFileListItemBeforeFirst.takeRetainedValue(), nil, nil, itemUrl, nil, nil).takeRetainedValue()
```

This line is the most important of the method. It adds the login item. It can't be removed otherwise we lose the start-at-login functionality. Or am I misunderstanding what you meant?

Reply:
> > I found that this wasn't building for me on both master, and this pr master on Xcode 13.1:
> 
> 
> 
> Do you know why there was a change (from compiling to no-longer compiling)? What's the new error with `nil` here?
> 

No clue but I suspect that apple forces updates on you when you update Xcode?
> 
> 
> > I don't see this line in LoginServiceKit, so I think it seems reasonable to remove this line @lwouis ?
> 
> 
> 
> Do you mean this line?
> 
> 
> 
> ```
> 
> let _ = LSSharedFileListInsertItemURL(loginItems, kLSSharedFileListItemBeforeFirst.takeRetainedValue(), nil, nil, itemUrl, nil, nil).takeRetainedValue()
> 
> ```
> 
> 
> 
> This line is the most important of the method. It adds the login item. It can't be removed otherwise we lose the start-at-login functionality. Or am I misunderstanding what you meant?

WELL haha, that's the line that crashes every single time for me, I wonder how LoginServiceKit and others do it

Reply:
I [shared above](https://github.com/lwouis/alt-tab-macos/pull/1094#discussion_r706216085) LoginServiceKit implementation, which is only around 50 lines, which are basically the same as what I did. They have some guard closes though, which mean they may just do nothing sometimes if some call goes wrong. Which means silently failing for the user.

I would like to see someone going to the bottom of this, and understanding why it's apparently crashing on BigSur. The APIs are still there AFAIK, so what is going on?

Reply:
Sorry but I didn't get the idea behind here, why should not use the LoginServiceKit, it does the same and it's fixed for the current macOS version? Why re implement the same?

Reply:
Hey @lwouis @ayroblu,

I replaced `LoginServiceKit` with the `LaunchAtLogin` swift package, which uses the Helper solution / `SMLoginItemSetEnabled` / `ServiceManagement` which works really well.

Please review this part again.

Reply:
I evaluated ServiceManagement during the initial implementation. I think of it as a [regression](https://github.com/sindresorhus/LaunchAtLogin#my-app-doesnt-show-up-in-system-preferences--users--groups--login-items) over the previous approaches. Currently, the user can manage their Login Items in all transparence. They could script or use apps to manage their Login Items, and AltTab is a part of that, making it accessible. With ServiceManagement, apps are no longer part of the Login Items standard approach, and users can't manage "login at launch" in the standard central place.

There are also other possible regressions. I know that the current implem is smart enough to [handle multiple versions](https://github.com/sindresorhus/LaunchAtLogin#my-app-doesnt-launch-at-login-when-testing) for instance.

I also worry about notarization. I've had numerous problems with it in the past, when introducing new frameworks for example. You can't test for it in the PR as well; it would only fail during CD on `master` as this is the only time/place where we use my paid Apple Developer account to notarize the app.

I'd prefer staying with the old API implem which I have tested for many many hours, and I know handles exotic edge cases. It's also light (few lines of code vs embedding an helper-app). If the `LoginServiceKit` library truly worked, then it should be pretty easy to peruse its implementation and adjust the current implem here to work on macOS 11.

This would be a path of least risk/change, with no regressions in terms of UX, app size, rare use-cases, etc.

Reply:
I know, I saw the issue. And yes, I also know that what Apple offers is not ideal, not even for the user.

But unfortunately this is the only supported way to start an app automatically when logging in.

And yes, introducing more things might lead to more problems.

I had implemented the helper myself first, if that helps, I can undo that as well. However, this framework looks like it is exactly what you want.

Reply:
But in fact, all the apps are getting rid of `LSSharedFileListInsertItemURL`. Also Apple is supporting `SMLoginItemSetEnabled` for Sandboxed apps.